### DISCUSSION TITLE
syntax-definitions:  KDE language syntax definition files (XML), or an array of values?

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#5722] [Q&A]
### Description

I'm wondering how one is supposed to use the [syntax-definitions field](https://quarto.org/docs/reference/formats/pdf.html#code); according to the docs:

```
syntax-definitions 	
KDE language syntax definition files (XML)
```

but when I have this in my yaml,

```
format:
  html:
    theme: cosmo
    highlight-style: arrow
    syntax-definitions: julia.xml
```

I get an immediate type-check error from quarto:

```
Field "syntax-definitions" has value julia.xml, which must instead be an array of values, where each element must be a string
68:     highlight-style: arrow
69:     syntax-definitions: julia.xml
                           ~~~~~~~~~~
✖ The value julia.xml is of type string.
```

Is there an example using this `syntax-definitions` option? I'm trying to see if I can improve the look of syntax highlighting in black and white, at the moment the words set in bold don't seem to match the language very well.

**Comment by baptiste** [comment]
Here's a [julia.xml](https://gist.github.com/baptiste/6b8efd23b324399c06d607b12cacee2c) definition file that I got from the KDE website

**Comment by mcanouil** [comment]
This is an option passed through directly to Pandoc, see <https://pandoc.org/MANUAL.html#syntax-highlighting>.

**Reply by baptiste** [reply]
I wonder how one can pass it through quarto. I've tried `syntax-definitions: ["julia.xml"]`, which passes the parsing check, but it seems to hang forever when knitting.

**Reply by mcanouil** [reply]
Did you try your file directly through Pandoc? `quarto pandoc file.qmd --from markdown --to whatever -o output.extension`

**Reply by baptiste** [reply]
It's also not working, so I guess there's more than one problem:

```
quarto pandoc test-hi.qmd --from markdown --to html --syntax-definition julia.xml -o test-hi.html
```

with this file

````
---
title: "Untitled"
---


```julia
using Plots

plot(sin, 
     x->sin(2x), 
     0, 
     2π, 
     leg=false, 
     fill=(0,:lavender))
```

````

gives me a warning:

```
Error compiling regex \.\.\.?|[:$]=?|\.?(?:>>>=|<-->|//=|<<=|>>=|-->|<--|===|!==|>>>|[-+*/\\^÷%|&⊻]=|=>|&&|\|\|?|>=|<=|==|!=|<\||\|>|\+\+|//|<<|>>|<:|>:|[\Q=≔⩴≕~←→↔↚↛↞↠↢↣↦↤↮⇎⇍⇏⇐⇒⇔⇴⇶⇷⇸⇹⇺⇻⇼⇽⇾⇿⟵⟶⟷⟹⟺⟻⟼⟽⟾⟿⤀⤁⤂⤃⤄⤅⤆⤇⤌⤍⤎⤏⤐⤑⤔⤕⤖⤗⤘⤝⤞⤟⤠⥄⥅⥆⥇⥈⥊⥋⥎⥐⥒⥓⥖⥗⥚⥛⥞⥟⥢⥤⥦⥧⥨⥩⥪⥫⥬⥭⥰⧴⬱⬰⬲⬳⬴⬵⬶⬷⬸⬹⬺⬻⬼⬽⬾⬿⭀⭁⭂⭃⭄⭇⭈⭉⭊⭋⭌￩￫⇜⇝↜↝↩↪↫↬↼↽⇀⇁⇄⇆⇇⇉⇋⇌⇚⇛⇠⇢↷↶↺↻><≥≤≡≠≢∈∉∋∌⊆⊈⊂⊄⊊∝∊∍∥∦∷∺∻∽∾≁≃≂≄≅≆≇≈≉≊≋≌≍≎≐≑≒≓≖≗≘≙≚≛≜≝≞≟≣≦≧≨≩≪≫≬≭≮≯≰≱≲≳≴≵≶≷≸≹≺≻≼≽≾≿⊀⊁⊃⊅⊇⊉⊋⊏⊐⊑⊒⊜⊩⊬⊮⊰⊱⊲⊳⊴⊵⊶⊷⋍⋐⋑⋕⋖⋗⋘⋙⋚⋛⋜⋝⋞⋟⋠⋡⋢⋣⋤⋥⋦⋧⋨⋩⋪⋫⋬⋭⋲⋳⋴⋵⋶⋷⋸⋹⋺⋻⋼⋽⋾⋿⟈⟉⟒⦷⧀⧁⧡⧣⧤⧥⩦⩧⩪⩫⩬⩭⩮⩯⩰⩱⩲⩳⩵⩶⩷⩸⩹⩺⩻⩼⩽⩾⩿⪀⪁⪂⪃⪄⪅⪆⪇⪈⪉⪊⪋⪌⪍⪎⪏⪐⪑⪒⪓⪔⪕⪖⪗⪘⪙⪚⪛⪜⪝⪞⪟⪠⪡⪢⪣⪤⪥⪦⪧⪨⪩⪪⪫⪬⪭⪮⪯⪰⪱⪲⪳⪴⪵⪶⪷⪸⪹⪺⪻⪼⪽⪾⪿⫀⫁⫂⫃⫄⫅⫆⫇⫈⫉⫊⫋⫌⫍⫎⫏⫐⫑⫒⫓⫔⫕⫖⫗⫘⫙⫷⫸⫹⫺⊢⊣⟂⫪⫫…⁝⋮⋱⋰⋯+-−¦|⊕⊖⊞⊟∪∨⊔±∓∔∸≏⊎⊻⊽⋎⋓⧺⧻⨈⨢⨣⨤⨥⨦⨧⨨⨩⨪⨫⨬⨭⨮⨹⨺⩁⩂⩅⩊⩌⩏⩐⩒⩔⩖⩗⩛⩝⩡⩢⩣*/⌿÷%&··⋅∘×\∩∧⊗⊘⊙⊚⊛⊠⊡⊓∗∙∤⅋≀⊼⋄⋆⋇⋉⋊⋋⋌⋏⋒⟑⦸⦼⦾⦿⧶⧷⨇⨰⨱⨲⨳⨴⨵⨶⨷⨸⨻⨼⨽⩀⩃⩄⩋⩍⩎⩑⩓⩕⩘⩚⩜⩞⩟⩠⫛⊍▷⨝⟕⟖⟗⨟^↑↓⇵⟰⟱⤈⤉⤊⤋⤒⤓⥉⥌⥍⥏⥑⥔⥕⥘⥙⥜⥝⥠⥡⥣⥥⥮⥯￪￬\E]): Failed reading: parse error at byte position 18
```

which, to be fair, looks like a rather unorthodox regex ¯\_(ツ)_/¯

**Reply by mcanouil** [reply]
So it would seem to be a Pandoc issue rather than Quarto.
Maybe worth checking Pandoc repository for related issues, and possibly report this one (if you are sure your file is correct).

**Comment by cderv** [comment]
> I wonder how one can pass it through quarto. I've tried syntax-definitions: ["julia.xml"], which passes the parsing check, but it seems to hang forever when knitting.

@baptiste I can confirm this the way to pass it . Autocompletion should help you here. It leads me to this below, which is equivalent to the above: 

````yaml
format: 
  html:
    syntax-definitions: 
      - julia.xml
````

Though I get the same error as you regarding the regex. It seems like the regex is in Pandoc's bundled file (https://github.com/jgm/skylighting/blob/master/skylighting-core/xml/julia.xml) but when using the CLI flag, it creates the issue. 

It is worth reporting upstream to PANDOC IMO, as @mcanouil already advice.

**Reply by baptiste** [reply]
I was a bit surprised by the quarto side of things as well, because the documentation is quite terse and without an example one might assume that passing a xml filename should work. 

I'll try to ask on the Pandoc list; it's puzzling that it uses it fine internally but not when passed from the command line.
